File: day1/Main.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) {
		File file = new File("input.txt");
		ArrayList<Integer> l1 = new ArrayList<>();
		ArrayList<Integer> l2 = new ArrayList<>();

		try (Scanner sc = new Scanner(file)) {
			while (sc.hasNext()) {
				String arr[] = sc.nextLine().split("\s+");
				l1.add(Integer.parseInt(arr[0]));
				l2.add(Integer.parseInt(arr[1]));
			}
		} catch (Exception e) {
			System.out.println(e);
			return;
		}

		System.out.println(answer1(l1, l2));
		System.out.println(answer2(l1, l2));
	}

	private static int answer1(List<Integer> l1, List<Integer> l2) {
		Collections.sort(l1);
		Collections.sort(l2);

		int sum = 0;

		for (int i = 0; i < l1.size(); i++) {
			sum += Math.abs(l1.get(i) - l2.get(i));
		}

		return sum;
	}

	private static int answer2(List<Integer> l1, List<Integer> l2) {
		Set<Integer> s1 = new HashSet<>(l1);
		Map<Integer, Integer> frequency = new HashMap<>();
		l2.forEach(num -> {
			frequency.merge(num, 1, (f, s) -> f + s);
		});

		return s1.stream().reduce(0, (x, y) -> x + y * frequency.getOrDefault(y, 0), (x, y) -> x + y);
	}
}

File: day1/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

type void struct{}

var member void

func main() {
	bytes, err := os.ReadFile("input.txt")

	if err != nil {
		fmt.Println("Error occured, %+v", err)
		return
	}

	newlines := regexp.MustCompile("\\r?\\n")

	lines := newlines.Split(string(bytes), -1)
	l1, l2 := []int{}, []int{}
	var num int

	for _, line := range lines {
		arr := strings.Split(line, "   ")
		if len(arr) == 2 {
			num, _ = strconv.Atoi(arr[0])
			l1 = append(l1, num)

			num, _ = strconv.Atoi(arr[1])
			l2 = append(l2, num)
		}
	}

	fmt.Println(answer1(l1, l2))
	fmt.Println(answer2(l1, l2))
}

func answer1(l1, l2 []int) int {
	slices.Sort(l1)
	slices.Sort(l2)

	var sum int
	for idx := range l1 {
		sum += int(math.Abs(float64(l1[idx] - l2[idx])))
	}
	return sum
}

func answer2(l1, l2 []int) int {
	set := make(map[int]void)
	for _, num := range l1 {
		set[num] = member
	}

	occurrences := make(map[int]int)
	for _, num := range l2 {
		occurrences[num]++
	}

	var sum int
	for k := range set {
		sum += occurrences[k] * k
	}

	return sum
}
